cmd/k8s-nameserver/main.go

```diff
@@ -153,11 +153,36 @@ func (n *nameserver) handleFunc() func(w dns.ResponseWriter, r *dns.Msg) {
 				m.Answer = append(m.Answer, rr)
 			}
 		case dns.TypeAAAA:
-			// TODO (irbekrm): implement IPv6 support.
-			// Kubernetes distributions that I am most familiar with
-			// default to IPv4 for Pod CIDR ranges and often many cases don't
-			// support IPv6 at all, so this should not be crucial for now.
-			fallthrough
+			// TODO (irbekrm): add IPv6 support.
+			// The nameserver currently does not support IPv6
+			// (records are not being created for IPv6 Pod addresses).
+			// However, we can expect that some callers will
+			// nevertheless send AAAA queries.
+			// We have to return NOERROR if a query is received for
+			// an AAAA record for a DNS name that we have an A
+			// record for- else the caller might not follow with an
+			// A record query.
+			// https://github.com/tailscale/tailscale/issues/12321
+			// https://datatracker.ietf.org/doc/html/rfc4074
+			q := r.Question[0].Name
+			fqdn, err := dnsname.ToFQDN(q)
+			if err != nil {
+				m = r.SetRcodeFormatError(r)
+				return
+			}
+			// The only supported use of this nameserver is as a
+			// single source of truth for MagicDNS names by
+			// non-tailnet Kubernetes workloads.
+			m.Authoritative = true
```

Comment:
Sorry I haven't looked much at this command before, so I'm not familiar: how do we know for sure at this point that `fqdn` is a MagicDNS name at all? If users forcefully direct a query to us with `google.com`, will we authoritatively respond that it doesn't exist?

Reply:
So, in normal usage, a query for `google.com` would never hit this nameserver, the only supported mode of operation for it is as a [stub nameserver for `ts.net` DNS names](https://tailscale.com/kb/1236/kubernetes-operator#expose-a-tailnet-https-service-to-your-cluster-workloads).
But also, we only create a handler for `ts.net` DNS names [here](https://github.com/tailscale/tailscale/blob/main/cmd/k8s-nameserver/main.go#L85), so we won't return authoritative for `google.com`:

```
$ dig @10.80.12.15 google.com

; <<>> DiG 9.18.24 <<>> @10.80.12.15 google.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: REFUSED, id: 12199
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;google.com.                    IN      A

```

Reply:
Thanks for the explainer! That's super helpful background.